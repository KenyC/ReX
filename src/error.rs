//! Defines different error types related to various phases of rendering a formula.
//!   - [`FontError`] : errors that have to do with the font file provided (missing MATH table, no such glyph).
//!   - [`ParseError`] : syntax error in the formula provided (mismatching brackets, unknown command).
//!   - [`LayoutError`] : errors during the layout phase ; currently, these can only be font errors.

use crate::font::common::GlyphId;
use crate::parser::error::ParseError;
use std::fmt;

/// Result type for the [`LayoutError`]
pub type LayoutResult<T> = ::std::result::Result<T, LayoutError>;

/// Errors during the layout phase ; currently, these can only be font errors.
#[derive(Debug, Clone, PartialEq)]
pub enum LayoutError {
    /// a font error
    Font(FontError)
}

/// Errors having to do with font file provided
#[derive(Debug, Clone, PartialEq)]
pub enum FontError {
    /// The font does not contain a glyph for the given char.
    MissingGlyphCodepoint(char),
    /// The font does not contain a glyph with that id.
    MissingGlyphGID(GlyphId),
    /// The font lacks a MATH table.
    NoMATHTable,
}

impl From<FontError> for LayoutError {
    fn from(e: FontError) -> Self {
        LayoutError::Font(e)
    }
}

/// A generic error type covering any error that may happen during the process.
#[derive(Debug, Clone, PartialEq)]
pub enum Error {
    /// a parse error
    Parse(ParseError),
    /// a layout error (including font errors)
    Layout(LayoutError)
}
impl From<ParseError> for Error {
    fn from(e: ParseError) -> Self {
        Error::Parse(e)
    }
}
impl From<LayoutError> for Error {
    fn from(e:LayoutError) -> Self {
        Error::Layout(e)
    }
}


impl fmt::Display for FontError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        use self::FontError::*;
        match *self {
            MissingGlyphCodepoint(cp) =>
                write!(f, "missing glyph for codepoint'{}'", cp),
            MissingGlyphGID(gid) =>
                write!(f, "missing glyph with gid {}", Into::<u16>::into(gid)),
            NoMATHTable =>
                write!(f, "no MATH tables"),
        }
    }
}
