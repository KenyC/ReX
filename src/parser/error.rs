//! Errors in parsing

use std::fmt;

use super::{control_sequence::PrimitiveControlSequence, GroupKind};


/// Result type for the [`ParseError`]
pub type ParseResult<T> = ::std::result::Result<T, ParseError>;


/// Syntax error in the formula provided (mismatching brackets, unknown command)
#[derive(Debug, Clone, PartialEq)]
pub enum ParseError {
    /// The symbol is not one we have atom type info about.
    UnrecognizedSymbol(char),
    /// There is no primitive control sequence with this name
    UnrecognizedControlSequence(Box<str>),
    /// Unable to parse argument of `\color{..}` as a color
    /// Valid color tokens are:
    ///  - Ascii name for css color (ie: `red`).
    ///  - #RRGGBB (ie: `#ff0000` for red)
    ///  - #RRGGBBAA (ie: `#00000000` for transparent)
    ///  - `transparent`
    UnrecognizedColor(Box<str>),
    /// A custom macro is missing an argument
    MissingArgForMacro {
        expected : usize,
        got : usize,
    },
    /// The brackets used to enclose a macro's arguments were not matched
    UnmatchedBrackets,
    /// A group (e.g. `{..}`, `\begin{env}..\end{env}`, `&...&`) was ended but there is no correponding begin group
    UnexpectedEndGroup{ 
        expected: Box<[GroupKind]>,
        got:      GroupKind,
    },
    /// A token or group of token was expected but never came
    ExpectedToken,
    /// An argument of control sequence like `\begin{..}` or `\color{..}` must be a sequence of chars ; it cannot contain a command
    ExpectedChars,
    /// A primitive control sequence needs a group as argument but can't find one (e.g. `{\sqrt}+1`).
    MissingArgForCommand(Box<str>),
    /// `\begin{array}` must be followed by a group describing column format (e.g. `\begin{array}{cc}`)
    MissingColFormatForArrayEnvironment,
    /// After `^` and `_`, a group or a token must follow.
    MissingSubSuperScript,
    /// There either is more than one subscript or more than one superscript attached to the same node.
    TooManySubscriptsOrSuperscripts,
    /// The command `\rule` expects an argument of the form `1.3pt` (number followed by dimension). The dimension may not be anything but `em` or `pt` at the moment.
    UnrecognizedDimension(Box<str>),
    /// The string in `\begin{..}` or `\end{..}` is not a recognized environment. Cf [Environment] for the list of supported LaTeX environments.
    UnrecognizedEnvironmen(Box<str>),
    /// The argument of `\begin{array}{..}` is not of the correct form: 
    /// it can only contain the characters `c`, `l`, `r`, whitespaces, braces, `|`  or macros that ultimately expand to one of these.
    UnrecognizedArrayColumnFormat,
    /// The token immediately following `\left`, `\middle` and `\right` isn't a symbol
    ExpectedDelimiter,
    /// The token immediately following `\left` is not of atom type [`AtomType::Open`] or  [`AtomType::Fence`]
    ExpectedOpenDelimiter,
    /// The token immediately following `\middle` is not of atom type [`AtomType::Fence`]
    ExpectedMiddleDelimiter,
    /// The token immediately following `\middle` is not of atom type [`AtomType::Close`]
    ExpectedClosingDelimiter,
    /// The command `\limits` and `\nolimits` must be placed right after an operator (or a macro that expands into something that ends in an operator)
    LimitControlSequenceMustBeAfterOperator,
}


impl fmt::Display for ParseError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        use self::ParseError::*;
        todo!()
    }
}
