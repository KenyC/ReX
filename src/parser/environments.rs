use crate::parser::error::ParseError;
use crate::parser::{tokens_as_string, List};

use super::nodes::{Array, ArrayColumnAlign, ArrayColumnsFormatting, ArraySingleColumnFormatting};
use super::symbols::Symbol;
use super::{error::ParseResult, nodes::CellContent, textoken::TexToken, Parser};
use super::GroupKind;

/// An enumeration of recognized enviornmnets.
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum Environment {
    Array,
    Matrix,
    PMatrix,
    BMatrix,
    BbMatrix,
    VMatrix,
    VvMatrix,
}

impl Environment {
    pub fn from_name(name : &str) ->  Option<Self> {
        match name {
            "array"    => Some(Self::Array),
            "matrix"   => Some(Self::Matrix),
            "pmatrix"  => Some(Self::PMatrix),
            "bmatrix"  => Some(Self::BMatrix),
            "Bmatrix"  => Some(Self::BbMatrix),
            "vmatrix"  => Some(Self::VMatrix),
            "Vmatrix"  => Some(Self::VvMatrix),
            _ => None
        }
    }
}


impl<'a, I : Iterator<Item = TexToken<'a>>> Parser<'a, I> {
    pub fn parse_environment(&mut self, env : Environment) -> ParseResult<Array> {
        let mut col_format = None;

        if let Environment::Array = env {
            let tokens = self.token_iter
                .capture_group()
                .map_err(|e| match e {
                    ParseError::ExpectedToken => ParseError::MissingColFormatForArrayEnvironment,
                    _ => e,
                })?;
            col_format = Some(tokens_as_column_format(tokens.into_iter())?);
        }
        let rows = self.parse_array_body(env)?;

        let left_delimiter;
        let right_delimiter;

        match env {
            Environment::Array    => {
                left_delimiter  = None;
                right_delimiter = None;
            },
            Environment::Matrix   => {
                left_delimiter  = None;
                right_delimiter = None;
            },
            Environment::PMatrix  => {
                left_delimiter  = Some(Symbol {codepoint : '(', atom_type : unicode_math::AtomType::Inner});
                right_delimiter = Some(Symbol {codepoint : ')', atom_type : unicode_math::AtomType::Inner});
            },
            Environment::BMatrix  => {
                left_delimiter  = Some(Symbol {codepoint : '[', atom_type : unicode_math::AtomType::Inner});
                right_delimiter = Some(Symbol {codepoint : ']', atom_type : unicode_math::AtomType::Inner});
            },
            Environment::BbMatrix => {
                left_delimiter  = Some(Symbol {codepoint : '{', atom_type : unicode_math::AtomType::Inner});
                right_delimiter = Some(Symbol {codepoint : '}', atom_type : unicode_math::AtomType::Inner});
            },
            Environment::VMatrix  => {
                left_delimiter  = Some(Symbol {codepoint : '|', atom_type : unicode_math::AtomType::Inner});
                right_delimiter = Some(Symbol {codepoint : '|', atom_type : unicode_math::AtomType::Inner});
            },
            Environment::VvMatrix => {
                left_delimiter  = Some(Symbol {codepoint : '\u{2016}', atom_type : unicode_math::AtomType::Inner});
                right_delimiter = Some(Symbol {codepoint : '\u{2016}', atom_type : unicode_math::AtomType::Inner});
            },
        }

        let col_format = col_format.unwrap_or_else(|| {
            let n_cols = rows.last().map_or(0, |row| row.len());
            ArrayColumnsFormatting {
                n_vertical_bars_before: 0,
                columns: vec![ArraySingleColumnFormatting { 
                    alignment: ArrayColumnAlign::Centered, 
                    n_vertical_bars_after: 0, 
                }; n_cols],
            }
        });

        Ok(Array {
            col_format,
            rows,
            left_delimiter,
            right_delimiter,
        })
    }


    pub fn parse_array_body(&mut self, env : Environment) -> ParseResult<Vec<Vec<CellContent>>> {
        let mut to_return    = Vec::new();
        let mut current_line = Vec::new();

        while {
            let List {nodes, group} = self.parse_until_end_of_group()?;

            match group {
                GroupKind::Env(env_ended) if env == env_ended => {
                    if !current_line.is_empty() || !nodes.is_empty() {
                        current_line.push(nodes);
                        to_return.push(std::mem::take(&mut current_line));
                    }
                    false
                },
                GroupKind::Align => {
                    current_line.push(nodes);
                    true
                },
                GroupKind::NewLine => {
                    current_line.push(nodes);
                    to_return.push(std::mem::take(&mut current_line));
                    true
                },

                _ => return Err(ParseError::UnexpectedEndGroup { expected : vec![GroupKind::Align, GroupKind::NewLine, GroupKind::Env(env)].into_boxed_slice(), got : group }),
            }
        }
        {}


        Ok(to_return)
    }
}

fn tokens_as_column_format<'a, I : Iterator<Item = TexToken<'a>>>(iterator : I) -> ParseResult<ArrayColumnsFormatting> {
    let mut n_vertical_bars_before = 0;
    let mut current_vertical_bars = &mut n_vertical_bars_before;
    let mut columns = Vec::new();
    for token in iterator {
        match token {
              TexToken::Char(c@'c') 
            | TexToken::Char(c@'l') 
            | TexToken::Char(c@'r') 
            => {
                columns.push(ArraySingleColumnFormatting {
                    alignment: match c {
                        'c' => ArrayColumnAlign::Centered,
                        'l' => ArrayColumnAlign::Left,
                        'r' => ArrayColumnAlign::Right,
                        _   => unreachable!(), // This has already been ruled out in the previous match
                    },
                    n_vertical_bars_after: 0,
                });
                current_vertical_bars = &mut columns.last_mut().unwrap_or_else(|| unreachable!()).n_vertical_bars_after;
            },
            TexToken::Char('|') 
            => {
                *current_vertical_bars += 1;
            },
            TexToken::Char(_) => {
                return Err(ParseError::UnrecognizedArrayColumnFormat);
            }
            TexToken::WhiteSpace => (),

            TexToken::BeginGroup 
            | TexToken::EndGroup 
            | TexToken::ControlSequence(_) 
            | TexToken::Superscript 
            | TexToken::Prime { .. } 
            | TexToken::Alignment 
            | TexToken::Subscript => return Err(ParseError::UnrecognizedArrayColumnFormat),
        }
    }
    Ok(ArrayColumnsFormatting {
        columns,
        n_vertical_bars_before,
    })
}


#[cfg(test)]
mod tests {
    use crate::parser::textoken::TokenIterator;

    use super::*;

    #[test]
    fn parse_col_format() {
        let cols = vec![
            ("c", 
            ArrayColumnsFormatting {
                columns : vec![
                    ArraySingleColumnFormatting {alignment : ArrayColumnAlign::Centered, n_vertical_bars_after : 0}
                ], 
                n_vertical_bars_before: 0
            }),
            ("||l", 
            ArrayColumnsFormatting {
                columns : vec![
                    ArraySingleColumnFormatting {alignment : ArrayColumnAlign::Left, n_vertical_bars_after : 0},
                ], 
                n_vertical_bars_before: 2
            }),
            ("||c|l", 
            ArrayColumnsFormatting {
                columns : vec![
                    ArraySingleColumnFormatting {alignment : ArrayColumnAlign::Centered, n_vertical_bars_after : 1},
                    ArraySingleColumnFormatting {alignment : ArrayColumnAlign::Left,     n_vertical_bars_after : 0},
                ], 
                n_vertical_bars_before: 2
            }),
            ("|  |c l|l|", 
            ArrayColumnsFormatting {
                columns : vec![
                    ArraySingleColumnFormatting {alignment : ArrayColumnAlign::Centered, n_vertical_bars_after : 0},
                    ArraySingleColumnFormatting {alignment : ArrayColumnAlign::Left,     n_vertical_bars_after : 1},
                    ArraySingleColumnFormatting {alignment : ArrayColumnAlign::Left,     n_vertical_bars_after : 1},
                ], 
                n_vertical_bars_before: 2
            }),
            (" |  r| l|| | |r||  ", 
            ArrayColumnsFormatting {
                columns : vec![
                    ArraySingleColumnFormatting {alignment : ArrayColumnAlign::Right, n_vertical_bars_after : 1},
                    ArraySingleColumnFormatting {alignment : ArrayColumnAlign::Left,  n_vertical_bars_after : 4},
                    ArraySingleColumnFormatting {alignment : ArrayColumnAlign::Right, n_vertical_bars_after : 2},
                ], 
                n_vertical_bars_before: 1
            }),
        ];

        for (string, col_format) in cols {
            let mut token_iter = TokenIterator::new(string);

            assert_eq!(
                tokens_as_column_format(token_iter),
                Ok(col_format),
            );

        }
        
    }


    #[test]
    fn good_arrays() {
        let collection = crate::parser::macros::CommandCollection::default();
        let mut parser = Parser::new(&collection, r"1&2\\3&4\end{pmatrix}");
        let result = parser.parse_environment(Environment::PMatrix);
        result.unwrap();
    }
}