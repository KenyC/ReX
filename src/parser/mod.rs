//! Parses strings representing LateX formulas into [`ParseNode`]'s
//! 
//! The main function function of interest is [`engine::parse`]

pub mod nodes;
pub mod color;
pub mod symbols;
pub mod macros;
pub mod error;
pub mod environments;
mod textoken;
mod control_sequence;

use unicode_math::AtomType;

use crate::error::ParseResult;
use crate::font::style_symbol;
use crate::font::Style;
use crate::parser::control_sequence::parse_color;
use crate::parser::nodes::GenFraction;
use crate::parser::nodes::PlainText;
use crate::parser::textoken::InputProcessor;
use crate::parser::textoken::TexToken;
use crate::parser::control_sequence::PrimitiveControlSequence;

use self::environments::Environment;
use self::error::ParseError;
use self::macros::CommandCollection;
use self::macros::ExpandedTokenIter;
pub use self::nodes::ParseNode;
pub use self::nodes::is_symbol;
use self::symbols::Symbol;
use self::textoken::TokenIterator;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GroupKind {
    BraceGroup,
    Env(Environment),
    // a group ended by &
    Cell,
    // a group end by \\
    Line,
    // end of file
    EndOfInput,

}

struct List {
    nodes : Vec<ParseNode>,
    group : GroupKind
}


/// Contains the internal state of the TeX parser, what's left to parse, and has methods to parse various TeX construct.  
/// Holds a reference to `CommandCollection`, which holds the definition of custom TeX macros defined by the user.
/// When not using custom macros, the parser can be made `'static`.
pub struct Parser<'a, I : Iterator<Item = TexToken<'a>>> {
    token_iter : ExpandedTokenIter<'a, I>,
    current_style : Style,
}

impl<'a> Parser<'a, TokenIterator<'a>> {
    pub fn new<'command : 'a, 'input : 'a>(command_collection: & 'command CommandCollection, input: & 'input str) -> Self { 
        Self { 
            token_iter : ExpandedTokenIter::new(command_collection, TokenIterator::new(input)),
            current_style : Style::default()
        } 
    }
}

impl<'a, I : Iterator<Item = TexToken<'a>>> Parser<'a, I> {

    const EMPTY_COMMAND_COLLECTION : & 'static CommandCollection = &CommandCollection::new();

    pub fn from_iter<'command : 'a>(command_collection: & 'command CommandCollection, input: I) -> Self { 
        Self { 
            token_iter : ExpandedTokenIter::new(command_collection, input),
            current_style: Style::default(),
        } 
    }

    pub fn parse(&mut self) -> ParseResult<Vec<ParseNode>> {
        let List { nodes, group } = self.parse_until_end_of_group()?;
        if let GroupKind::EndOfInput = group 
        { Ok(nodes) }
        else 
        { Err(ParseError::UnexpectedEndGroup(group)) }
    }


    fn parse_until_end_of_group(&mut self) -> ParseResult<List> {
        let mut results = Vec::new();

        while let Some(token) = self.token_iter.next_token()? {
            match token {
                TexToken::Char('^') | TexToken::Char('_')  => {

                },
                _ if token.is_end_group() => {
                    return Ok(List { nodes: results, group: GroupKind::BraceGroup });
                },
                TexToken::Char(codepoint) => {
                    let atom_type = codepoint_atom_type(codepoint).ok_or_else(|| ParseError::UnrecognizedSymbol(codepoint))?;
                    let codepoint = style_symbol(codepoint, self.current_style);
                    results.push(ParseNode::Symbol(Symbol { codepoint, atom_type }));
                },
                // Here we deal with "primitive" control sequences, not macros
                TexToken::ControlSequence(control_sequence_name) => {
                    let command = 
                        PrimitiveControlSequence::from_name(control_sequence_name)
                        .ok_or_else(|| ParseError::UnrecognizedControlSequence(control_sequence_name.to_string().into_boxed_str()))?
                    ;
                    use PrimitiveControlSequence::*;
                    match command {
                        Radical => {
                            let inner = self.parse_required_control_seq_argument_as_nodes(control_sequence_name)?;
                            results.push(ParseNode::Radical(nodes::Radical { inner, }));
                        },
                        Rule => todo!(),
                        Color => {
                            let color_name_group = self.token_iter.capture_group()?;
                            let color = parse_color(color_name_group.into_iter())?;
                            let inner = self.parse_required_control_seq_argument_as_nodes(control_sequence_name)?;
                            results.push(ParseNode::Color(nodes::Color {
                                color,
                                inner,
                            }));
                        },
                        ColorLit(color) => {
                            let inner = self.parse_required_control_seq_argument_as_nodes(control_sequence_name)?;
                            results.push(ParseNode::Color(nodes::Color {
                                color,
                                inner,
                            }));
                        },
                        Fraction(left_delimiter, right_delimiter, bar_thickness, style) => {
                            let numerator   = self.parse_required_control_seq_argument_as_nodes(control_sequence_name)?;
                            let denominator = self.parse_required_control_seq_argument_as_nodes(control_sequence_name)?;

                            results.push(ParseNode::GenFraction(GenFraction {
                                numerator, denominator,
                                left_delimiter, right_delimiter,
                                bar_thickness, style,
                            }));
                        },
                        DelimiterSize(_, _) => todo!(),
                        Kerning(space) => {
                            results.push(ParseNode::Kerning(space))
                        },
                        StyleCommand(_) => todo!(),
                        AtomChange(_) => todo!(),
                        TextOperator(_, _) => todo!(),
                        SubStack(_) => todo!(),
                        Text => {
                            let text_group = self.token_iter.capture_group()?;
                            let text = tokens_as_string(text_group.into_iter())?;
                            results.push(ParseNode::PlainText(PlainText {
                                text,
                            }));
                        },
                        BeginEnv => {
                            let env_name_group = self.token_iter.capture_group()?;
                            let env_name = tokens_as_string(env_name_group.into_iter())?;
                            todo!()
                        },
                        EndEnv => todo!(),
                        SymbolCommand(symbol) => {
                            let Symbol { codepoint, atom_type } = symbol;
                            let codepoint = style_symbol(codepoint, self.current_style);
                            results.push(ParseNode::Symbol(Symbol { codepoint, atom_type, }));
                        },
                    }
                },
            }
        }

        Ok(List { nodes: results, group: GroupKind::EndOfInput })
    }

    fn parse_required_control_seq_argument_as_nodes(&mut self, control_seq_name : &str) -> ParseResult<Vec<ParseNode>> {
        self.parse_required_argument_as_nodes()
            .map_err(|e| match e {
                ParseError::ExpectedToken => ParseError::MissingArgForCommand(Box::from(control_seq_name)),
                e => e,
            })
    }

    fn parse_required_argument_as_nodes(&mut self) -> ParseResult<Vec<ParseNode>> {
        let group = self.token_iter.capture_group()?;

        // Normally all tokens are already expanded after `capture_group`
        // There is no need to have further expansions
        let mut forked_parser = Parser::from_iter(Self::EMPTY_COMMAND_COLLECTION, group.into_iter());
        forked_parser.current_style = self.current_style;

        let List { nodes, group } = forked_parser.parse_until_end_of_group()?;

        if group != GroupKind::EndOfInput {
            return Err(ParseError::UnexpectedEndGroup(group));
        }
        
        Ok(nodes)
    }
}

fn tokens_as_string<'a, I : Iterator<Item = TexToken<'a>>>(iterator : I) -> ParseResult<String> {
    let mut to_return = String::new();
    for token in iterator {
        match token {
            TexToken::Char(c) => to_return.push(c),
            TexToken::ControlSequence(_) => return Err(ParseError::ExpectedChars),
        }
    }
    Ok(to_return)
}


/// This function is the API entry point for parsing tex.
pub fn parse(input: &str) -> ParseResult<Vec<ParseNode>> {
    parse_with_custom_commands(input, &CommandCollection::default())
}


pub fn parse_with_custom_commands<'a>(input: & 'a str, custom_commands : &CommandCollection) -> ParseResult<Vec<ParseNode>> {
    Parser::new(custom_commands, input).parse()
}







/// Helper function for determining an atomtype based on a given codepoint.
/// This is primarily used for characters while processing, so may give false
/// negatives when used for other things.
fn codepoint_atom_type(codepoint: char) -> Option<AtomType> {
    Some(match codepoint {
             'a' ..= 'z' | 'A' ..= 'Z' | '0' ..= '9' | 'Α' ..= 'Ω' | 'α' ..= 'ω' => AtomType::Alpha,
             '*' | '+' | '-' => AtomType::Binary,
             '[' | '(' => AtomType::Open,
             ']' | ')' | '?' | '!' => AtomType::Close,
             '=' | '<' | '>' | ':' => AtomType::Relation,
             ',' | ';' => AtomType::Punctuation,
             '|' => AtomType::Fence,
             '/' | '@' | '.' | '"' => AtomType::Alpha,
             _ => return None,
         })
}



#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn snapshot_symbols() {
        insta::assert_debug_snapshot!(parse("1"));
        insta::assert_debug_snapshot!(parse("a"));
        insta::assert_debug_snapshot!(parse("+"));
        insta::assert_debug_snapshot!(parse(r"\mathrm A"));
        insta::assert_debug_snapshot!(parse(r"\mathfrak A"));
        insta::assert_debug_snapshot!(parse(r"\alpha"));
        // should object to cyrillic characters
        insta::assert_debug_snapshot!(parse(r"Ж"));
    }

    #[test]
    fn snapshot_frac() {
        insta::assert_debug_snapshot!(parse(r"\frac 12"));
        insta::assert_debug_snapshot!(parse(r"\frac{1+0} {2+2}"));
        insta::assert_debug_snapshot!(parse(r"\frac \left(1\right)2"));
        insta::assert_debug_snapshot!(parse(r"\frac\alpha\beta"));
    }

    #[test]
    fn snapshot_radicals() {
        // success
        insta::assert_debug_snapshot!(parse(r"\sqrt{x}"));
        insta::assert_debug_snapshot!(parse(r"\sqrt2"));
        insta::assert_debug_snapshot!(parse(r"\sqrt\alpha"));
        insta::assert_debug_snapshot!(parse(r"1^\sqrt2"));
        insta::assert_debug_snapshot!(parse(r"\alpha_\sqrt{1+2}"));
        insta::assert_debug_snapshot!(parse(r"\sqrt\sqrt2"));
        insta::assert_debug_snapshot!(parse(r"\sqrt2_3" ));
        insta::assert_debug_snapshot!(parse(r"\sqrt{2_3}"));

        // fail
        insta::assert_debug_snapshot!(parse(r"\sqrt" ));
        insta::assert_debug_snapshot!(parse(r"\sqrt_2" ));
        insta::assert_debug_snapshot!(parse(r"\sqrt^2"));
    }


    #[test]
    fn snapshot_scripts() {
        insta::assert_debug_snapshot!(parse(r"1_2"));
        insta::assert_debug_snapshot!(parse(r"1_2^3"));
        insta::assert_debug_snapshot!(parse(r"1^3_2"));
        insta::assert_debug_snapshot!(parse(r"1^\alpha"));
        insta::assert_debug_snapshot!(parse(r"1^2^3"));
        insta::assert_debug_snapshot!(parse(r"1^{2^3}"));
        insta::assert_debug_snapshot!(parse(r"{a^b}_c"));
        insta::assert_debug_snapshot!(parse(r"1_{1+1}^{2+1}"));
    }


    #[test]
    fn snapshot_delimited() {
        // success
        insta::assert_debug_snapshot!(parse(r"\left(\right)"));
        insta::assert_debug_snapshot!(parse(r"\left(\right."));
        insta::assert_debug_snapshot!(parse(r"\left(\alpha\right)"));
        insta::assert_debug_snapshot!(parse(r"\left(\alpha+1\right)"));
        insta::assert_debug_snapshot!(parse(r"\left(1\middle|2\right)"));
        insta::assert_debug_snapshot!(parse(r"\left(1\middle|2\middle|3\right)"));
        insta::assert_debug_snapshot!(parse(r"\left\lBrack{}x\right\rBrack"));

        // fail
        insta::assert_debug_snapshot!(parse(r"\left(1\middle|"));
        insta::assert_debug_snapshot!(parse(r"\right(1+1"));
        insta::assert_debug_snapshot!(parse(r"\left)1+1\right)"));
    }


    #[test]
    fn snapshot_array() {
        insta::assert_debug_snapshot!(parse(r"\begin{array}{c}\end{array}"));
        insta::assert_debug_snapshot!(parse(r"\begin{array}{c}1\\2\end{array}"));
        insta::assert_debug_snapshot!(parse(r"\begin{array}{c}1\\\end{array}"));
        insta::assert_debug_snapshot!(parse(r"\begin{pmatrix}1&2\\3&4\end{pmatrix}"));
        insta::assert_debug_snapshot!(parse(r"\begin{array}{c|l}1&\alpha\\2&\frac12\end{array}"));
        insta::assert_debug_snapshot!(parse(r"\begin{array}{cc}1 \\ 2"));
    }

    #[ignore = "unsupported as of yet"]
    #[test]
    fn snapshot_rule() {
        insta::assert_debug_snapshot!(parse(r"\rule{1cm}{3pt}"));
        insta::assert_debug_snapshot!(parse(r"\rule{4pt}{5px}"));
    }

    #[test]
    fn snapshot_plain_text() {
        insta::assert_debug_snapshot!(parse(r"\text{abc}"));
        insta::assert_debug_snapshot!(parse(r"\text{abc}def"));
        insta::assert_debug_snapshot!(parse(r"\text{\{\}1}1}"));
        insta::assert_debug_snapshot!(parse(r"\text{}}"));
    }

    #[test]
    fn snapshot_color() {
        // success
        insta::assert_debug_snapshot!(parse(r"\color{cyan}{1+1}"));
        insta::assert_debug_snapshot!(parse(r"\color{red}{1+1}"));
        insta::assert_debug_snapshot!(parse(r"\red{1}"));
        insta::assert_debug_snapshot!(parse(r"\blue{1}"));
        insta::assert_debug_snapshot!(parse(r"\gray{1}"));
        insta::assert_debug_snapshot!(parse(r"\color{chartreuse}\alpha"));
        insta::assert_debug_snapshot!(parse(r"\color{chocolate}\alpha"));

        // fail
        insta::assert_debug_snapshot!(parse(r"\color{bred}{1+1}"));
        insta::assert_debug_snapshot!(parse(r"\color{bred}1"));
        insta::assert_debug_snapshot!(parse(r"\color red{1}"));
    }

    #[test]
    fn snapshot_style() {
        // success
        insta::assert_debug_snapshot!(parse(r"1\scriptstyle2"));
        insta::assert_debug_snapshot!(parse(r"{1\scriptstyle}2"));
        insta::assert_debug_snapshot!(parse(r"1\textstyle2"));
        insta::assert_debug_snapshot!(parse(r"1\sqrt{\displaystyle s}1"));
    }


    #[test]
    fn snapshot_atom_change() {
        // success
        insta::assert_debug_snapshot!(parse(r"1\mathrel{R}2"));
        insta::assert_debug_snapshot!(parse(r"1\mathrel{\frac{1}{2}} 2"));
        insta::assert_debug_snapshot!(parse(r"\mathop{1}2"));
    }


    #[test]
    fn snapshot_text_operators() {
        // success
        insta::assert_debug_snapshot!(parse(r"\sin 1"));
        insta::assert_debug_snapshot!(parse(r"\log (42 + 1)"));
        insta::assert_debug_snapshot!(parse(r"\sin(a + b) = \sin a \cos b + \cos b \sin a"));
        insta::assert_debug_snapshot!(parse(r"\det_{B} M"));
        insta::assert_debug_snapshot!(parse(r"\lim_{h \to 0 } \frac{f(x+h)-f(x)}{h}"));
    }


    #[test]
    fn snapshot_spacing() {
        // success
        insta::assert_debug_snapshot!(parse(r"1\!2"));
        insta::assert_debug_snapshot!(parse(r"2\quad 3"));
        insta::assert_debug_snapshot!(parse(r"2\quad3"));
        insta::assert_debug_snapshot!(parse(r"5\,2"));
        insta::assert_debug_snapshot!(parse(r"5\;2"));
        insta::assert_debug_snapshot!(parse(r"5\:2"));
        insta::assert_debug_snapshot!(parse(r"1\qquad{}33"));

        // failure
        insta::assert_debug_snapshot!(parse(r"1\33"));
    }

    #[test]
    fn snapshot_delimiter() {
        // success
        insta::assert_debug_snapshot!(parse(r"\biggl("));
        insta::assert_debug_snapshot!(parse(r"\bigr]"));
        insta::assert_debug_snapshot!(parse(r"\Bigl\langle"));
        insta::assert_debug_snapshot!(parse(r"\Biggr|"));
        insta::assert_debug_snapshot!(parse(r"\Bigl\lBrack"));

        // failure
        insta::assert_debug_snapshot!(parse(r"\bigr\lBrack"));
        insta::assert_debug_snapshot!(parse(r"\Bigl\rangle"));
    }

    #[test]
    fn snapshot_substack() {
        // success
        insta::assert_debug_snapshot!(parse(r"\substack{   1 \\ 2}"));
        insta::assert_debug_snapshot!(parse(r"\substack{ 1 \\ \frac{7}8 \\ 4}"));
        insta::assert_debug_snapshot!(parse(r"\begin{array}{c}\substack{1 \\ \frac{7}8 \\ 4} \\ 5 \end{array}"));
        insta::assert_debug_snapshot!(parse(r"\substack{1 \\}"));
        insta::assert_debug_snapshot!(parse(r"1 \substack{}"));

        // failure
        insta::assert_debug_snapshot!(parse(r"\substack{ 1 \\ 2}\\"));
        insta::assert_debug_snapshot!(parse(r"\substack \alpha \\ 1"));
        insta::assert_debug_snapshot!(parse(r"\substack{ 1 \\ 1"));
    }

    #[test]
    fn snapshot_style_change() {
        // success
        insta::assert_debug_snapshot!(parse(r"1\scriptstyle 2"));
        insta::assert_debug_snapshot!(parse(r"1\scriptstyle2\textstyle1+1"));
        insta::assert_debug_snapshot!(parse(r"1{\scriptstyle2\textstyle1}+1"));
        insta::assert_debug_snapshot!(parse(r"\frac{22\scriptscriptstyle22}2"));
    }


    #[test]
    fn snapshot_primes() {
        insta::assert_debug_snapshot!(parse("a'"));
        insta::assert_debug_snapshot!(parse("a''"));
        insta::assert_debug_snapshot!(parse("a'''"));
        insta::assert_debug_snapshot!(parse("a''''"));
        insta::assert_debug_snapshot!(parse("'a"));
        insta::assert_debug_snapshot!(parse(r"\sqrt'"));
    }
}
