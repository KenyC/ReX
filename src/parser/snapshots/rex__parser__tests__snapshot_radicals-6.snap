---
source: src/parser/mod.rs
expression: "parse(r\"\\sqrt\\sqrt2\")"
---
Err(
    MissingArgForCommand(
        "sqrt",
    ),
)
